maskrcnn_benchmark/modeling/rpn/retinanet/inference.py

```diff
@@ -0,0 +1,225 @@
+
+
+
+class RetinaNetPostProcessor(torch.nn.Module):
+    """
+    Performs post-processing on the outputs of the RetinaNet boxes.
+    This is only used in the testing.
+    """
+    def __init__(
+        self,
+        pre_nms_thresh,
+        pre_nms_top_n,
+        nms_thresh,
+        fpn_post_nms_top_n,
+        min_size,
+        box_coder=None,
+    ):
+        """
+        Arguments:
+            pre_nms_thresh (float)
+            pre_nms_top_n (int)
+            nms_thresh (float)
+            fpn_post_nms_top_n (int)
+            min_size (int)
+            box_coder (BoxCoder)
+        """
+        super(RetinaNetPostProcessor, self).__init__()
+        self.pre_nms_thresh = pre_nms_thresh
+        self.pre_nms_top_n = pre_nms_top_n
+        self.nms_thresh = nms_thresh
+        self.fpn_post_nms_top_n = fpn_post_nms_top_n
+        self.min_size = min_size
+
+        if box_coder is None:
+            box_coder = BoxCoder(weights=(10., 10., 5., 5.))
+        self.box_coder = box_coder
+
+    def forward_for_single_feature_map(
+            self, anchors, box_cls, box_regression, pre_nms_thresh=0.05):
+        """
+        Arguments:
+            anchors: list[BoxList]
+            box_cls: tensor of size N, A * C, H, W
+            box_regression: tensor of size N, A * 4, H, W
+        """
+        device = box_cls.device
+        N, _, H, W = box_cls.shape
+        A = int(box_regression.size(1) / 4)
+        C = int(box_cls.size(1) / A)
+
+        # put in the same format as anchors
+        box_cls = box_cls.view(N, -1, C, H, W).permute(0, 3, 4, 1, 2)
+        box_cls = box_cls.reshape(N, -1, C)
+        box_cls = box_cls.sigmoid()
+
+        box_regression = box_regression.view(N, -1, 4, H, W)
+        box_regression = box_regression.permute(0, 3, 4, 1, 2)
+        box_regression = box_regression.reshape(N, -1, 4)
+
+        num_anchors = A * H * W
+
+        results = [[] for _ in range(N)]
+        candidate_inds = box_cls > pre_nms_thresh
+        if candidate_inds.sum().item() == 0:
+            empty_boxlists = []
+            for a in anchors:
+                empty_boxlist = BoxList(torch.Tensor(0, 4).to(device), a.size)
+                empty_boxlist.add_field(
+                    "labels", torch.LongTensor([]).to(device))
+                empty_boxlist.add_field(
+                    "scores", torch.Tensor([]).to(device))
+                empty_boxlists.append(empty_boxlist)
+            return empty_boxlists
+
+
+        pre_nms_top_n = candidate_inds.view(N, -1).sum(1)
+        pre_nms_top_n = pre_nms_top_n.clamp(max=self.pre_nms_top_n)
+
+        for batch_idx, (per_box_cls, per_box_regression, per_pre_nms_top_n, \
+        per_candidate_inds, per_anchors) in enumerate(zip(
+            box_cls,
+            box_regression,
+            pre_nms_top_n,
+            candidate_inds,
+            anchors)):
+
+            # Sort and select TopN
+            per_box_cls = per_box_cls[per_candidate_inds]
```

Comment:
Can't the whole implementation that is in here accept empty tensors?
In this case, do we really need to special-case if there are no detection, as in line 72 here?

Reply:
I am not sure about this part. This version definitely is cleaner. But the early return is important to the speed performance. 
Test : RetinaNet - Resnet50 on 1080 TI) 
Previous version : 123.9 ms
This version : 149.4 ms

Reply:
That is surprising.
All the operations that happen in this for loop are on empty tensors, so there should be just a skip in the C++ side that would happen.
I'd like to understand where this difference comes from: isn't it just because of the sync point in line 94 due to the `.item()`?

Also, how often do we have images with no detections? For a trained model, that shouldn't happen often, should it?

Reply:
According to my experience, there are 2~3 images in COCO **minival** could cause the problems. (no detections with confidence >= 0.05)  Besides, I think no detection should be the normal case in inference. 
This problem also happens in Detectron. The Detectron solution is to use threshold 0.0 in the last year(P7). I also test this solution but it slows down more. (around 160~170 ms)

Reply:
So, if I understand it correctly, there are (approximately) two images over 5000 on coco for which the speed would go from 123ms to 149ms, is that right? That doesn't seem like a problem if we consider the whole evaluation time, is that right?

Reply:
No. Maybe I didn't express correctly.  In RetinaNet, there are 5 layers(P3-P7) used for prediction. It's very common that no detections with confidence more than 0.05 in some layers. I think that's the reason why early return can reduce the inference time. In some rare cases( 2 images over 5000 on COCO), there is no detection >0.05 in all prediction layers.

Reply:
Oh, now I got it, thanks!
Still, I'm quite surprised by this slowdown. I'd like to understand where it comes from.
It could be due to advanced indexing, but it has been recently optimized in PyTorch master.

Reply:
I think that we can obtain a speed improvement here without much effort:
we move
```python
per_candidate_nonzeros = per_candidate_inds.nonzero()
```
before this line, and pass `per_candidate_nonzeros` instead 
```python
per_box_cls = per_box_cls[per_box_loc, per_class]
```
How does this change the runtime, could you check?

Reply:
Okay. This step does not change current inference time much. The inference time is around 0.1477 second/per image.

Reply:
It's a bit manual, but maybe you could save a boolean flag for `is_empty = candidate_inds.sum().item() == 0` and then skip only parts of the for loop to see what exactly is causing the slowdown (i.e. does it happen when you  call the `self.box_code.decode`, does it happen if you make everything below `if per_candidate_inds_sum().item() > ...` only execute if you have `not is_empty`. That way it would be easier to pinpoint which bits the network (wastefully) spends its time in.
I imagine lots of syncing when looking at the indexing (and the .item() anyways), so I wonder whether it might even help to do the entire indexing on CPU and transfer the tensors to CPU before the for loop.